framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,563 @@
+            result.append(' ');
+        }
+        result.append(fieldName);
+        return result.toString();
+    }
+
+    /**
+     * Writes out an import statement for each annotation used in an {@link AScene}.
+     *
+     * <p>{@code DefCollector} is a facility in the Annotation File Utilities for determining which
+     * annotations are used in a given AScene. Here, we use that construct to write out the proper
+     * import statements into a stub file.
+     */
+    private static class ImportDefWriter extends DefCollector {
+
+        /** The writer onto which to write the import statements. */
+        private final PrintWriter printWriter;
+
+        /**
+         * Constructs a new ImportDefWriter, which will run on the given AScene when its {@code
+         * visit} method is called.
+         *
+         * @param scene the scene whose imported annotations should be printed
+         * @param printWriter the writer onto which to write the import statements
+         * @throws DefException if the DefCollector does not succeed
+         */
+        ImportDefWriter(ASceneWrapper scene, PrintWriter printWriter) throws DefException {
+            super(scene.getAScene());
+            this.printWriter = printWriter;
+        }
+
+        /**
+         * Write an import statement for a given AnnotationDef. This is only called once per
+         * annotation used in the scene.
+         *
+         * @param d the annotation definition to print an import for
+         */
+        @Override
+        protected void visitAnnotationDef(AnnotationDef d) {
+            if (!isInternalJDKAnnotation(d.name)) {
+                printWriter.println("import " + d.name + ";");
+            }
+        }
+    }
+
+    /**
+     * Do not print internal JDK annotations, which are the only annotations that include a '+'
+     *
+     * @param annotationName the name of the annotation
+     * @return true iff this is an internal JDK annotation that should not be printed
+     */
+    private static boolean isInternalJDKAnnotation(String annotationName) {
+        return annotationName.contains("+");
+    }
+
+    /**
+     * For a given class, this prints the hierarchy of outer classes, and returns the number of
+     * curly braces to close with. The classes are printed with appropriate opening curly braces, in
+     * standard Java style. This routine does not attempt to indent them correctly.
+     *
+     * <p>When an inner class is present in an AScene, its name is something like "Outer.Inner".
+     * Writing a stub file with that name would be useless to the stub parser, which expects inner
+     * classes to be properly nested.
+     *
+     * @param basename the binary name of the class with the package part stripped
+     * @param aClass the AClass for {@code classname}
+     * @param printWriter the writer where the class definition should be printed
+     * @return the number of outer classes within which this class is nested
```

Comment:
Shouldn't this return the total number of classes printed rather than the number minus one?  That leads to an odd `1 +` elsewhere.